pkg/types/rfc3161/v0.0.1/entry.go

```diff
@@ -0,0 +1,189 @@
+	"errors"
+	"fmt"
+
+	"github.com/sigstore/rekor/pkg/types/rfc3161"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/sassoftware/relic/lib/pkcs9"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/types"
+)
+
+const (
+	APIVERSION = "0.0.1"
+)
+
+func init() {
+	if err := rfc3161.VersionMap.SetEntryFactory(APIVERSION, NewEntry); err != nil {
+		log.Logger.Panic(err)
+	}
+}
+
+type V001Entry struct {
+	Rfc3161Obj models.Rfc3161V001Schema
+}
+
+func (v V001Entry) APIVersion() string {
+	return APIVERSION
+}
+
+func NewEntry() types.EntryImpl {
+	return &V001Entry{}
+}
+
+func (v V001Entry) IndexKeys() []string {
+	var result []string
+
+	if v.HasExternalEntities() {
+		if err := v.FetchExternalEntities(context.Background()); err != nil {
+			log.Logger.Error(err)
+			return result
+		}
+	}
+
+	str := v.Rfc3161Obj.Tsr.Content.String()
+	tb, err := base64.StdEncoding.DecodeString(str)
+	if err != nil {
+		log.Logger.Warn(err)
+	} else {
+		h := sha256.Sum256(tb)
+		hx := hex.EncodeToString(h[:])
+
+		payloadKey := "sha256:" + hx
+		result = append(result, payloadKey)
+	}
+
+	return result
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	rfc3161Resp, ok := pe.(*models.Rfc3161)
+	if !ok {
+		return errors.New("cannot unmarshal non Rfc3161 v0.0.1 type")
+	}
+
+	if err := types.DecodeEntry(rfc3161Resp.Spec, &v.Rfc3161Obj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := v.Rfc3161Obj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (v V001Entry) HasExternalEntities() bool {
+	return false
+}
+
+func (v *V001Entry) FetchExternalEntities(ctx context.Context) error {
+	return nil
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	if err := v.FetchExternalEntities(ctx); err != nil {
+		return nil, err
+	}
+
+	if v.Rfc3161Obj.Tsr == nil || v.Rfc3161Obj.Tsr.Content == nil {
+		return nil, fmt.Errorf("tsr content is nil")
+	}
+
+	content := *v.Rfc3161Obj.Tsr.Content
+
+	// validate the content
+	b, err := base64.StdEncoding.DecodeString(content.String())
+	if err != nil {
+		return nil, err
+	}
+	var tsr pkcs9.TimeStampResp
+	_, err = asn1.Unmarshal(b, &tsr)
+	if err != nil {
+		return nil, err
+	}
+	if tsr.Status.Status != pkcs9.StatusGranted && tsr.Status.Status != pkcs9.StatusGrantedWithMods {
+		return nil, fmt.Errorf("Tsr status not granted: %v", tsr.Status.Status)
+	}
+	if !tsr.TimeStampToken.ContentType.Equal(asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 7, 2}) {
+		return nil, fmt.Errorf("Tsr wrong content type: %v", tsr.TimeStampToken.ContentType)
+	}
+	sig, err := tsr.TimeStampToken.Content.Verify(nil, false)
+	if err != nil {
+		return nil, fmt.Errorf("Tsr verification error: %v", err)
+	}
+
+	cert := strfmt.Base64(sig.Certificate.Raw)
+
+	canonicalEntry := models.Rfc3161V001Schema{
+		Tsr: &models.Rfc3161V001SchemaTsr{
+			Content: &content,
```

Comment:
should we be persisting the entire TSR content into the log (this deviates from the other types, where we do not store content itself)? I could see extracting the message imprint (content digest and algorithm)... @dlorenc what are your thoughts here?

Reply:
I **think** it would be safe to store the whole thing, but I honestly don't understand the format well enough. @asraa what do you think? It's definitely simplest to just store it all...

Reply:
i don't think we should just store the message imprint, since the TSR contains that *plus* the time added (and two of the same messages with different timestamps would be equivalent)

if you wanted a minimal portion of the "signed" TSR content, that would be the TSTInfo (that's the thing the server generates with the matching message imprint and time). Even that, that's not the actual bytes the server signs, but it should be unique if considered with the public key.

Reply:
the actual signed bytes would be the AuthenticatedAttributes. https://github.com/sigstore/rekor/blob/c4b1a3941339dd67f6419ba368142b70ecfc5750/pkg/util/rfc3161.go#L230

that would be unique, it contains the hash of the TSTInfo and the hash of the signing cert.

Reply:
I don't have a strong opinion here, I can change to store anything. Should it just be TSTInfo and the Cert/PublicKey?

Reply:
Let's store the entire thing.

Reply:
after thinking about this as well, I think we should store the entire thing.